File: src/main/java/com/yupi/airouter/controller/RechargeController.java
package com.yupi.airouter.controller;

import com.mybatisflex.core.paginate.Page;
import com.stripe.model.checkout.Session;
import com.yupi.airouter.annotation.AuthCheck;
import com.yupi.airouter.common.BaseResponse;
import com.yupi.airouter.common.ResultUtils;
import com.yupi.airouter.config.StripeConfig;
import com.yupi.airouter.constant.UserConstant;
import com.yupi.airouter.exception.BusinessException;
import com.yupi.airouter.exception.ErrorCode;
import com.yupi.airouter.model.entity.RechargeRecord;
import com.yupi.airouter.model.entity.User;
import com.yupi.airouter.service.RechargeService;
import com.yupi.airouter.service.StripePaymentService;
import com.yupi.airouter.service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.annotation.Resource;
import jakarta.servlet.http.HttpServletRequest;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.web.bind.annotation.*;

import java.io.Serializable;
import java.math.BigDecimal;

/**
 * 充值控制器
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
@RestController
@RequestMapping("/recharge")
@Slf4j
public class RechargeController {

    @Resource
    private RechargeService rechargeService;

    @Resource
    private StripePaymentService stripePaymentService;

    @Resource
    private UserService userService;

    @Resource
    private StripeConfig stripeConfig;

    /**
     * 创建充值订单（Stripe）
     */
    @PostMapping("/stripe/create")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "创建Stripe充值订单")
    public BaseResponse<CreateRechargeResponse> createStripeRecharge(
            @RequestBody CreateRechargeRequest request,
            HttpServletRequest httpRequest) {

        if (request.getAmount() == null || request.getAmount().compareTo(BigDecimal.ZERO) <= 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "充值金额必须大于0");
        }

        // 设置最小充值金额1元，最大10000元
        if (request.getAmount().compareTo(BigDecimal.ONE) < 0 ||
            request.getAmount().compareTo(new BigDecimal("10000")) > 0) {
            throw new BusinessException(ErrorCode.PARAMS_ERROR, "充值金额必须在1-10000元之间");
        }

        User loginUser = userService.getLoginUser(httpRequest);

        // 从配置获取回调URL
        String successUrl = stripeConfig.getSuccessUrl();
        String cancelUrl = stripeConfig.getCancelUrl();

        // 创建 Stripe 支付会话
        Session session = stripePaymentService.createCheckoutSession(
                loginUser.getId(),
                request.getAmount(),
                successUrl,
                cancelUrl
        );

        CreateRechargeResponse response = new CreateRechargeResponse();
        response.setCheckoutUrl(session.getUrl());
        response.setSessionId(session.getId());

        return ResultUtils.success(response);
    }

    /**
     * 充值成功回调
     */
    @GetMapping("/stripe/success")
    @Operation(summary = "Stripe充值成功回调")
    public BaseResponse<String> stripeSuccess(@RequestParam("session_id") String sessionId) {
        log.info("收到Stripe支付成功回调，SessionID: {}", sessionId);

        boolean success = stripePaymentService.handlePaymentSuccess(sessionId);
        if (success) {
            return ResultUtils.success("充值成功！");
        } else {
            throw new BusinessException(ErrorCode.OPERATION_ERROR, "充值处理失败");
        }
    }

    /**
     * 充值取消回调
     */
    @GetMapping("/stripe/cancel")
    @Operation(summary = "Stripe充值取消回调")
    public BaseResponse<String> stripeCancel() {
        log.info("用户取消了Stripe支付");
        return ResultUtils.success("您取消了充值");
    }

    /**
     * 获取我的充值记录（分页）
     */
    @GetMapping("/list/my")
    @AuthCheck(mustRole = UserConstant.DEFAULT_ROLE)
    @Operation(summary = "获取我的充值记录")
    public BaseResponse<Page<RechargeRecord>> getMyRechargeRecords(
            @RequestParam(defaultValue = "1") int pageNum,
            @RequestParam(defaultValue = "10") int pageSize,
            HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Page<RechargeRecord> page = rechargeService.listUserRechargeRecords(
                loginUser.getId(), pageNum, pageSize);
        return ResultUtils.success(page);
    }

    /**
     * 创建充值请求
     */
    @Data
    public static class CreateRechargeRequest {
        private BigDecimal amount; // 充值金额（元）
    }

    /**
     * 创建充值响应
     */
    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    public static class CreateRechargeResponse implements Serializable {
        private String checkoutUrl; // Stripe支付页面URL
        private String sessionId; // Stripe会话ID
    }
}

File: go-backend/internal/controller/recharge_controller.go
package controller

import (
	"log"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/config"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/service"
)

const (
	minRechargeAmount = 1.0
	maxRechargeAmount = 10000.0
)

type RechargeController struct {
	rechargeService      *service.RechargeService
	stripePaymentService *service.StripePaymentService
	userService          *service.UserService
	cfg                  *config.Config
}

func NewRechargeController(
	rechargeService *service.RechargeService,
	stripePaymentService *service.StripePaymentService,
	userService *service.UserService,
	cfg *config.Config,
) *RechargeController {
	return &RechargeController{
		rechargeService:      rechargeService,
		stripePaymentService: stripePaymentService,
		userService:          userService,
		cfg:                  cfg,
	}
}

func (r *RechargeController) CreateStripeRecharge(c *gin.Context) {
	var request dto.CreateRechargeRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		common.Error(c, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	if request.Amount < minRechargeAmount || request.Amount > maxRechargeAmount {
		common.Error(c, errno.ParamsError.Code, "充值金额必须在1-10000元之间")
		return
	}
	loginUser, err := r.userService.GetLoginUser(c)
	if err != nil {
		if bizErr, ok := errno.AsBusinessError(err); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	sess, err := r.stripePaymentService.CreateCheckoutSession(
		loginUser.ID,
		request.Amount,
		r.cfg.StripeSuccessURL,
		r.cfg.StripeCancelURL,
	)
	if err != nil {
		if bizErr, ok := errno.AsBusinessError(err); ok {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	common.Success(c, dto.CreateRechargeResponse{
		CheckoutURL: sess.URL,
		SessionID:   sess.ID,
	})
}

func (r *RechargeController) StripeSuccess(c *gin.Context) {
	sessionID := c.Query("session_id")
	if sessionID == "" {
		common.Error(c, errno.ParamsError.Code, errno.ParamsError.Message)
		return
	}
	ok, err := r.stripePaymentService.HandlePaymentSuccess(sessionID)
	if err != nil {
		if bizErr, hasBizErr := errno.AsBusinessError(err); hasBizErr {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	if !ok {
		common.Error(c, errno.OperationError.Code, "充值处理失败")
		return
	}
	common.Success(c, "充值成功！")
}

func (r *RechargeController) StripeCancel(c *gin.Context) {
	common.Success(c, "您取消了充值")
}

func (r *RechargeController) GetMyRechargeRecords(c *gin.Context) {
	loginUser, err := r.userService.GetLoginUser(c)
	if err != nil {
		if bizErr, hasBizErr := errno.AsBusinessError(err); hasBizErr {
			common.Error(c, bizErr.Code, bizErr.Message)
			return
		}
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	pageNum := int64(1)
	pageSize := int64(10)
	if raw := c.Query("pageNum"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageNum = parsed
		}
	}
	if raw := c.Query("pageSize"); raw != "" {
		if parsed, parseErr := strconv.ParseInt(raw, 10, 64); parseErr == nil && parsed > 0 {
			pageSize = parsed
		}
	}
	page, serviceErr := r.rechargeService.ListUserRechargeRecords(loginUser.ID, pageNum, pageSize)
	if serviceErr != nil {
		log.Printf("list recharge records failed: userId=%d err=%v", loginUser.ID, serviceErr)
		common.Error(c, errno.SystemError.Code, errno.SystemError.Message)
		return
	}
	common.Success(c, page)
}
